src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UIMetadataImportService.java

```diff
@@ -1,172 +1,242 @@
  */
 @Component
 public class UIMetadataImportService {
 
+	private static final Logger logger = LoggerFactory.getLogger(UIMetadataImportService.class);
+	private final MessageSource messageSource;
+	private final ProjectService projectService;
+	private final SampleService sampleService;
+	private final MetadataTemplateService metadataTemplateService;
+	private final ProjectControllerUtils projectControllerUtils;
+	private final UIMetadataFileImportService metadataFileImportService;
+
 	@Autowired
-	public UIMetadataImportService() {
+	public UIMetadataImportService(MessageSource messageSource, ProjectService projectService,
+			SampleService sampleService, MetadataTemplateService metadataTemplateService,
+			ProjectControllerUtils projectControllerUtils, UIMetadataFileImportService metadataFileImportService) {
+		this.messageSource = messageSource;
+		this.projectService = projectService;
+		this.sampleService = sampleService;
+		this.metadataTemplateService = metadataTemplateService;
+		this.projectControllerUtils = projectControllerUtils;
+		this.metadataFileImportService = metadataFileImportService;
 	}
 
 	/**
-	 * Parse metadata from an csv file.
+	 * Upload CSV or Excel file containing sample metadata and extract the headers.  The file is stored in the session until
+	 * the column that corresponds to a {@link Sample} identifier has been sent.
 	 *
-	 * @param inputStream The inputStream of the csv file.
-	 * @return {@link SampleMetadataStorage} contains the metadata from file.
-	 * @throws IOException thrown if the extension does not exist.
+	 * @param session   {@link HttpSession}
+	 * @param projectId {@link Long} identifier for the current {@link Project}
+	 * @param file      {@link MultipartFile} The csv or excel file containing the metadata.
+	 * @return {@link Map} of headers and rows from the csv or excel file for the user to select the header corresponding the
+	 * {@link Sample} identifier.
 	 */
-	public SampleMetadataStorage parseCSV(InputStream inputStream) throws IOException {
+	public SampleMetadataStorage createProjectSampleMetadata(HttpSession session, long projectId, MultipartFile file) {
+		// We want to return a list of the table headers back to the UI.
 		SampleMetadataStorage storage = new SampleMetadataStorage();
-
-		CSVParser parser = CSVParser.parse(inputStream, StandardCharsets.UTF_8,
-				CSVFormat.RFC4180.withFirstRecordAsHeader()
-						.withTrim()
-						.withIgnoreEmptyLines());
-		List<Map<String, String>> rows = new ArrayList<>();
-
-		// save headers
-		Map<String, Integer> headers_set = parser.getHeaderMap();
-		List<String> headers_list = new ArrayList<>(headers_set.keySet());
-		storage.saveHeaders(headers_list);
-
-		// save data
-		for (CSVRecord row : parser) {
-			Map<String, String> rowMap = new HashMap<>();
-			for (String key : row.toMap()
-					.keySet()) {
-				String value = row.toMap()
-						.get(key);
-				rowMap.put(key, value);
+		try {
```

Comment:
The try/catch stuff here can be simplified to be autoclosing so we don't have to worry about closing the input stream manually by doing this:


		try(InputStream is = file.getInputStream()) {
			// code here
		} catch (FileNotFoundException e) {
			logger.debug("No file found for uploading an excel file of metadata.");
		} catch (IOException e) {
			logger.error("Error opening file" + file.getOriginalFilename());
		}

You will no longer need this code in the try block as the above try statement will open the inputstream to the file

	byte[] byteArr = file.getBytes();
	InputStream inputStream = new ByteArrayInputStream(byteArr);

You will just need to verify that it still reads the file correctly

Reply:
Please see 2590dc306287b11feb77d4315902f9bbb876ee88.